net: throw error if port does not exist in options

yanivfriedensohn opened the pull request:
Throw error ERR_PROPERTY_NOT_IN_OBJECT if the port property does not
exist in the options object argument when calling server.listen().

Fixes: https://github.com/nodejs/node/issues/16712

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

jasnell commented:
Semver-major because of the addition of a new throw.

@nodejs/tsc ... PTAL. I'm not yet convinced this is definitely something we should do.

Trott commented:
Hi, @yanivfriedensohn and welcome! Thanks for the PR.

I think this error is misleading, no? `port` is not required. One of either `port` or `path` is required. Might it not be better to improve the message displayed for the existing `ERR_INVALID_OPT_VALUE` error?

yanivfriedensohn commented:
Hi @Trott, based on your suggestions, I added an additional error message argument to `ERR_INVALID_OPT_VALUE` and removed `ERR_PROPERTY_NOT_IN_OBJECT`.

joyeecheung commented:
How does this fix https://github.com/nodejs/node/issues/16712 ? It requested

> APIs that accept a host should respect their port number to avoid confusion and inconsistency. I would also suggest that server.listen() should only support hostname instead of host, leaving parsing hosts to userland... but that would be a breaking change (and should probably be a separate issue).

joyeecheung commented:
Note that `server.listen()` works like `server.listen(0)` so to say that `port` or `path` must be specified is not entirely accurate..

yanivfriedensohn commented:
@maclover7 I added tests for the additional error message.

@joyeecheung this fixes:
> The { host : 'localhost' } error message would be good to fix but the other "expected behaviors" could potentially create (possibly severe) ecosystem fallout for not enough value to take the risk.

The message mentions that `port` or `path` are required in `options`. Should I rephrase the message to make it more clear?

joyeecheung commented:
@yanivfriedensohn In that case please do not put `Fixes` in the PR/commit message as landing it would close that issue automatically even though it's not fully fixed.

joyeecheung commented on lib/net.js:
```diff
@@ -1496,7 +1496,15 @@ Server.prototype.listen = function(...args) {
 
-  throw new ERR_INVALID_OPT_VALUE('options', util.inspect(options));
+  let additionalMessage;
+  if (typeof options === 'object' &&
```
`typeof options === 'object'` here is always true since it's normalized before

joyeecheung commented on lib/net.js:
```diff
@@ -1496,7 +1496,15 @@ Server.prototype.listen = function(...args) {
+    additionalMessage = '"port" or "path" must be specified in options.';
+  }
+
+  throw new ERR_INVALID_OPT_VALUE('options', util.inspect(options),
```
This error message will not be accurate since `options` is an argument, not an option value - the user does not pass `{ options: { ... } }` to this method. `ERR_INVALID_ARG_VALUE` may be more accurate.

yanivfriedensohn commented:
@joyeecheung following your suggestion, I used `ERR_INVALID_ARG_VALUE` instead of `ERR_INVALID_OPT_VALUE` and reverted the change to `ERR_INVALID_OPT_VALUE`.

Trott commented:
Tests have been added so I've removed the `in progress` label. /cc @maclover7 Feel free to add it back if it still applies.

yanivfriedensohn commented:
@joyeecheung is this ready for CI?

yanivfriedensohn commented:
@Trott issue https://github.com/nodejs/node/issues/18435 seems to be related to the failed test.

Trott commented:
Resume Build: https://ci.nodejs.org/job/node-test-pull-request/16705/

yanivfriedensohn commented:
@Trott is this `author ready`?
